### ISSUE TITLE
Docker image out of date?

### ISSUE THREAD

**Issue by cboettig** [dsten/jupyterhub-deploy#2]
@ryanlovett Trying to make sense of what is going on here, looks like the https://github.com/dsten/jupyterhub-deploy/blob/master/roles/singleuser/files/Dockerfile image is based on `jupyter/systemuser`, which is based on `ipython/scipystack` which has been in a broken build state for three months and sounds like it may not be the right base image any more anyhow? (e.g. https://github.com/ipython/docker-notebook/issues/49)   

Tracing from the current `FROM` directives suggests that [this](https://github.com/dsten/jupyterhub-deploy/blob/master/roles/singleuser/files/Dockerfile) would inherit from `jupyter/notebook` but given the broken state of `ipython/scipystack` builds this seems not to be the case.  In any event, I can't find `jupyter` in the path on that image, which is a bit confusing to me.  Thanks for setting me straight.

**Comment by ryanlovett** [comment]
jupyter is not on the path because ipython/scipystack is FROM
ipython/ipython:3.x.

Thanks for pointing us at jupyter/docker-stack. Next semester we'll
probably look at jupyter/datascience-notebook.


Ryan

**Comment by cboettig** [comment]
very cool. I noticed `ipython/ipython` is now just `FROM jupyter/notebook`, but I suppose that just applies to `:latest` and not `3.x` which I had overlooked.  

I should have given more context: I'm teaching a DS8 connector in the Spring and was trying to get a better feel for what the DS8 Foundations course had under the hood for the Jupyter notebooks environment, and if/how I would extend it for my connector.  Clearly it isn't the Dockerfile I mentioned above then.  Can you point me to the right setup?  

the stack at jupyter/datascience-notebook is pretty nice IMHO; might have some inspiration for BCE.

**Comment by ryanlovett** [comment]
dsten/jupyterhub-deploy is the correct tree. It goes:

https://github.com/dsten/jupyterhub-deploy/blob/master/roles/singleuser/files/Dockerfile
  https://hub.docker.com/r/jupyter/systemuser/~/dockerfile/
    https://hub.docker.com/r/ipython/scipystack/~/dockerfile/
      ipython/ipython:3.x

I think you should definitely have input on the platform for your
connector. There have been a couple of isolated requests from connector
instructors (39c: "make pivot tables more excel-like" , 39h: "make python2
kernel available"). There may be more but those two happen to be using Stat
department labs so there are opportunities to chat them up. We should get
everyone's feedback later in the semester to help shape the stack in the
Spring.

(fyi, @papajohn)


Ryan

**Comment by ryanlovett** [comment]
https://groups.google.com/forum/#!topic/jupyter/tfVaVCQaNXM


Ryan

**Comment by SamLau95** [comment]
Is this still an open issue? Also see #4.

**Comment by cboettig** [comment]
er, I dunno.  Is there a docker image I can pull and get the same environment I see at data8.berkeley.edu?

**Comment by ryanlovett** [comment]
Yes,
https://github.com/data-8/jupyterhub-deploy/blob/master/roles/singleuser/files/Dockerfile
.


Ryan

**Comment by cboettig** [comment]
@ryanlovett thanks!  Any reason not to have this on an autobuild on Docker Hub?

**Comment by ryanlovett** [comment]
No, in fact that has always been the plan. I've been wanting the hub to
trigger builds when new datascience releases are pushed out.


Ryan

**Comment by cboettig** [comment]
You know Docker Hub has an API for that right?  I'm not sure what the `datascience` release workflow looks like, but if it's a scripted task you can just add a POST request that will trigger the image to rebuild on the hub.  (In the [rocker](https://github.com/rocker-org) project we do something similar to get rstudio-daily builds).  

Just having automated builds enabled for commits to this repository would be convenient though too.

**Comment by cboettig** [comment]
@ryanlovett apologies for my ignorance here: looks like this image is designed to be run automatically by juypterhub rather than actually launched by a human a la the jupyter/datascience-notebook and other images in docker-stacks.  e.g. doesn't look like one can do:

```
git clone https://github.com/data-8/jupyterhub-deploy
docker build -t data-8/jupyter roles/singleuser/files
docker run -P data-8/jupyter
```

since that just complains: 

```
Creating user  ()
useradd: invalid user ID '-s'
sudo: unknown user: /opt/conda/bin/python3
sudo: unable to initialize policy plugin
```

So how is one supposed to launch this environment (outside of a full jupyterhub environment)?

**Comment by ryanlovett** [comment]
You can run with "docker run -it data-8/jupyter /opt/conda/bin/python" (or /bin/bash) though this is not the easiest way of testing the notebooks. It is usually invoked with the /srv/singleuser/systemuser.sh wrapper which will generate the errors you cited when invoked without the Hub environment. It needs to be replaced with a different bootstrap script to run as a local web service on a standalone machine.

**Comment by ryanlovett** [comment]
Actually, the bootstrap script is start-notebook.sh via the inherited minimal-notebook container. Looks like it needs to be massaged a bit to get it to run simply.

**Comment by cboettig** [comment]
Thanks @ryanlovett !  Running with `start-notebook.sh` as the command seems to work rather nicely for me, as far as I can tell from a quick try at any rate. (Good to know this is just the file inherited from `minimal-notebook`, looks like the configuration options should then work as documented there?)

It's probably a silly idea, but I've had some students asking about this stuff and how to run Jupyter themselves outside of our server environment, so I'm thinking I will give this a whirl via docker toolbox and see what happens.  Might be a disaster but getting everyone going with local anaconda installs isn't exactly a cakewalk, so I'm curious to see how the dockerized approach compares.  

Think you could flip the switch on docker-hub to build this Dockerfile there?

**Comment by ryanlovett** [comment]
Updating my comment from early in this thread, the complete stack is now:
```
https://github.com/data-8/jupyterhub-deploy/blob/master/roles/singleuser/files/Dockerfile
  ⌞ https://hub.docker.com/r/jupyter/systemuser/~/dockerfile/
    ⌞ https://hub.docker.com/r/jupyter/singleuser/~/dockerfile/
      ⌞ https://github.com/jupyter/docker-stacks/blob/master/scipy-notebook/Dockerfile
        ⌞ https://github.com/jupyter/docker-stacks/blob/master/minimal-notebook/Dockerfile
          ⌞ debian
```

I'd be interested in how it works for your students since 1) it gives people at least a fallback option for running the environment, and 2) you know docker and can provide immediate support for your students. :)  I really, really, *really* want to make the course site(s) are more robust so that we don't have to support individual environments on personal devices, but I also understand the need and don't want to be in the way.

I built an older version a few weeks ago at https://hub.docker.com/r/ucbds/systemuser and I've been meaning to hook it to the deployment Dockerfile. Just haven't had time yet.

**Comment by cboettig** [comment]
@ryanlovett Thanks! I think it went remarkably smoothly today.  We started from scratch -- students downloaded and installed docker toolbox on the laptops while I introduced the basic idea of Docker and why we were going to try it.  The ~ 16 macs had no trouble, but less smooth on the ~ 4 Windows machines.  One worked fine out of the box, two needed virtualization enabled in BIOS, and the last would only give me the "something is wrong" error when trying to run docker quickstart terminal.

20 students pulling the ~ 5GB docker image down at once was obviously a bit slow, but I was pretty impressed, maybe 30 minutes before everyone had it.  (I had copies on a few thumb drives as a backup, but the download times were looking fast enough not to bother with it). I copied your Dockerfile to my GitHub account so that I could create a Hub build to reduce the number of steps required.  My rough notes I used for the lesson are here: http://berkeley.carlboettiger.info/espm-88b/tools/docker

We just covered very basic docker start commands for launching Jupyter and for linking working directories.  Still not sure how many docker commands a student would have to learn long term to make this to work.  I also didn't try to get the image working with Kitematic yet (I don't quite get how the command-line arguments for linking volumes or ports work in Kitematic, I'm usually just using Docker on Linux), but it would be nice to have a fully-GUI version.  

Not sure how this compares to the ease of the Anaconda approach overall.  Of course having a more identical platform helps.  Several students had already created local instances for Anconda and Jupyter.  I did have trouble myself setting up Anconda on a Mac when I tested it (see https://github.com/data-8/datascience/issues/212 ) but that might be just me.  

Some students did ask about other official images, so I pointed out the official python and jupyter images on Docker Hub.   It would be great to have a data-8 org account on Docker Hub.  

Cheers.
